reviewer/reviewer.go

```diff
@@ -0,0 +1,264 @@
+		}
+
+		if resp.NextPage == 0 {
+			break
+		}
+		opt.Page = resp.NextPage
+	}
+
+	return allUsers, nil
+}
+
+// select n random values from the list of users
+func selectRandomUsers(n int, users []string) []string {
+	generated := map[int]struct{}{}
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	var selections []string
+	if n == 0 {
+		return selections
+	}
+	if n >= len(users) {
+		return users
+	}
+
+	for i := 0; i < n; i++ {
+		for {
+			i := r.Intn(len(users))
+			if _, ok := generated[i]; !ok {
+				generated[i] = struct{}{}
+				selections = append(selections, users[i])
+				break
+			}
+		}
+	}
+	return selections
+}
+
+func shoveIntoMap(m map[string]struct{}, u []string) map[string]struct{} {
+	for _, n := range u {
+		m[n] = struct{}{}
+	}
+	return m
+}
+
+func FindRandomRequesters(ctx context.Context, prctx pull.Context, result common.Result, client *github.Client) ([]string, error) {
+	logger := zerolog.Ctx(ctx)
+	pendingLeafNodes := findLeafChildren(result)
+	var requestedUsers []string
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	logger.Debug().Msgf("Collecting reviewers for %d pending leaf nodes", len(pendingLeafNodes))
+
+	for _, child := range pendingLeafNodes {
+		allUsers := make(map[string]struct{})
+		allUsers = shoveIntoMap(allUsers, child.RequestedUsers)
+
+		if len(child.RequestedTeams) > 0 {
+			randomTeam := child.RequestedTeams[r.Intn(len(child.RequestedTeams))]
+			teamInfo := strings.Split(randomTeam, "/")
+			team, _, err := client.Teams.GetTeamBySlug(ctx, teamInfo[0], teamInfo[1])
+			if err != nil {
+				logger.Debug().Err(err).Msgf("Unable to get member listing for team %s", randomTeam)
+				//return nil, errors.Wrapf(err, "Unable to get information for team %s", randomTeam)
+			} else {
+				teamMembers, err := listAllTeamMembers(ctx, client, team)
+				if err != nil {
+					logger.Debug().Err(err).Msgf("Unable to get member listing for team %s", randomTeam)
+					//return nil, errors.Wrapf(err, "Unable to get member listing for team %s", randomTeam)
+				} else {
+					allUsers = shoveIntoMap(allUsers, teamMembers)
+				}
+			}
+		}
+
+		if len(child.RequestedOrganizations) > 0 {
+			randomOrg := child.RequestedOrganizations[r.Intn(len(child.RequestedOrganizations))]
+			orgMembers, err := listAllOrgMembers(ctx, client, randomOrg)
+			if err != nil {
+				return nil, errors.Wrapf(err, "Unable to get member listing for org %s", randomOrg)
+			}
+			allUsers = shoveIntoMap(allUsers, orgMembers)
+		}
+
+		allCollaborators, err := listAllCollaborators(ctx, client, prctx.RepositoryOwner(), prctx.RepositoryName())
+		if err != nil {
+			return nil, errors.Wrap(err, "Unable to get collaborators")
+		}
+		collaboratorPermissions := make(map[string]string)
+
+		if child.RequestedWriteCollaborators || child.RequestedAdmins {
+			for _, c := range allCollaborators {
+				perm, _, err := client.Repositories.GetPermissionLevel(ctx, prctx.RepositoryOwner(), prctx.RepositoryName(), c)
```

Comment:
I see what you mean about this being expensive. Here's a place where using GraphQL can help, as the [`RepositoryCollaboratorEdge`](https://developer.github.com/v4/object/repositorycollaboratoredge/) type includes the permission for the collaborator. That will let you list collaborators and their permissions in one operation.

Reply:
I've done this now and it seems much nicer.